tests/saml/metadata/federations/test_validator.py

```diff
@@ -0,0 +1,134 @@
+from nose.tools import assert_raises
+from parameterized import parameterized
+
+import tests.saml.fixtures
+from api.saml.metadata.federations import incommon
+from api.saml.metadata.federations.model import SAMLFederation
+from api.saml.metadata.federations.validator import (
+    SAMLFederatedMetadataExpirationValidator,
+    SAMLFederatedMetadataValidationError,
+    SAMLMetadataSignatureValidator,
+)
+
+
+class TestSAMLFederatedMetadataExpirationValidator(object):
+    @parameterized.expand(
+        [
+            (
+                "incorrect_xml",
+                datetime.datetime.utcnow(),
+                tests.saml.fixtures.INCORRECT_XML,
+                SAMLFederatedMetadataValidationError,
+            ),
+            (
+                "without_valid_until_attribute",
+                datetime.datetime.utcnow(),
+                tests.saml.fixtures.FEDERATED_METADATA_WITHOUT_VALID_UNTIL_ATTRIBUTE,
+                SAMLFederatedMetadataValidationError,
+            ),
+            (
+                "with_expired_valid_until_attribute",
+                tests.saml.fixtures.FEDERATED_METADATA_VALID_UNTIL
+                + SAMLFederatedMetadataExpirationValidator.MAX_CLOCK_SKEW
+                + datetime.timedelta(minutes=1),
+                tests.saml.fixtures.FEDERATED_METADATA_WITH_VALID_UNTIL_ATTRIBUTE,
+                SAMLFederatedMetadataValidationError,
+            ),
+            (
+                "with_valid_until_attribute_too_far_in_the_future",
+                tests.saml.fixtures.FEDERATED_METADATA_VALID_UNTIL
+                - SAMLFederatedMetadataExpirationValidator.MAX_VALID_TIME
+                - datetime.timedelta(minutes=1),
+                tests.saml.fixtures.FEDERATED_METADATA_WITH_VALID_UNTIL_ATTRIBUTE,
+                SAMLFederatedMetadataValidationError,
+            ),
+            (
+                "with_valid_until_attribute_less_than_current_time_and_less_than_max_clock_skew",
+                tests.saml.fixtures.FEDERATED_METADATA_VALID_UNTIL
+                + SAMLFederatedMetadataExpirationValidator.MAX_CLOCK_SKEW,
+                tests.saml.fixtures.FEDERATED_METADATA_WITH_VALID_UNTIL_ATTRIBUTE,
+                None,
+            ),
+            (
+                "with_valid_until_attribute_greater_than_current_time_and_less_than_max_valid_time",
+                tests.saml.fixtures.FEDERATED_METADATA_VALID_UNTIL
+                - SAMLFederatedMetadataExpirationValidator.MAX_VALID_TIME
+                + datetime.timedelta(minutes=1),
+                tests.saml.fixtures.FEDERATED_METADATA_WITH_VALID_UNTIL_ATTRIBUTE,
+                None,
+            ),
+            (
+                "with_real_incommon_metadata",
+                datetime.datetime(2020, 11, 26, 14, 32, 42),
+                open(
+                    os.path.join(
+                        os.path.dirname(os.path.abspath(__file__)),
+                        "../../../files/saml/incommon-metadata-idp-only.xml",
```

Comment:
Is it necessary to include this entire 38M document to do a good test? If the answer is yes then that's fine; I ran the test to check on timing and it only added about 1 second to the test time. But it's the biggest file in the repo so I want to see whether there's a way to run this test with a subset of the metadata, or whether it's better to make sure we handle every real case.

Reply:
Unfortunately, we can't just rip a snippet out of this document. The document contains a signed signature and the snippet won't pass validation of the signature because snippet's hash would be different.
However, there is a workaround: we can create a new pair of keys, sign the snippet using the private keys and use the public one to validate it.
But I just wanted to make sure that the real data will be processed correctly.

Reply:
This makes sense.